pkg/metrics/metrics.go

```diff
@@ -0,0 +1,190 @@
+	registry                     *prometheus.Registry
+	CertificateRequests          *prometheus.CounterVec
+	CertificateExpiryTimeSeconds *prometheus.GaugeVec
+}
+
+func New() *Metrics {
+
+	router := mux.NewRouter()
+
+	// Create server and register prometheus metrics handler
+	s := &Metrics{
+		Server: http.Server{
+			Addr:           prometheusMetricsServerAddress,
+			ReadTimeout:    prometheusMetricsServerReadTimeout,
+			WriteTimeout:   prometheusMetricsServerWriteTimeout,
+			MaxHeaderBytes: prometheusMetricsServerMaxHeaderBytes,
+			Handler:        router,
+		},
+		registry: prometheus.NewRegistry(),
+		CertificateRequests: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: Namespace,
+				Name:      "certificate_requests",
+				Help:      "Number of certificate requests",
+			},
+			[]string{"name", "namespace", "issuer_name", "issuer_kind", "issuer_type", "operation", "result"},
+		),
+		CertificateExpiryTimeSeconds: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: Namespace,
+				Name:      "certificate_expiry_time_seconds",
+				Help:      "Number of seconds after January 1, 1970 UTC that the certificate will expire",
+			},
+			[]string{"name", "namespace"},
+		),
+	}
+
+	s.registry.MustRegister(s.CertificateRequests)
+	s.registry.MustRegister(s.CertificateExpiryTimeSeconds)
+
+	router.Handle("/metrics", promhttp.HandlerFor(s.registry, promhttp.HandlerOpts{}))
+
+	return s
+}
+
+func (m *Metrics) waitShutdown(stopCh <-chan struct{}) {
+	<-stopCh
+	glog.Info("Stopping Prometheus metrics server...")
+
+	ctx, cancel := context.WithTimeout(context.Background(), prometheusMetricsServerShutdownTimeout)
+	defer cancel()
+
+	if err := m.Shutdown(ctx); err != nil {
+		glog.Errorf("Prometheus metrics server shutdown error: %v", err)
+		return
+	}
+
+	glog.Info("Prometheus metrics server gracefully stopped")
+}
+
+func (m *Metrics) Start(stopCh <-chan struct{}) {
+	go func() {
+
+		glog.Infof("Listening on http://%s", m.Addr)
+		if err := m.ListenAndServe(); err != nil {
+			glog.Errorf("Error running prometheus metrics server: %s", err.Error())
+			return
+		}
+
+		glog.Infof("Prometheus metrics server exited")
+
+	}()
+
+	m.waitShutdown(stopCh)
+}
+
+// handleCertificate increments the certificate request count
+// for a particular type and result
+func (m *Metrics) handleCertificate(crt *v1alpha1.Certificate, i v1alpha1.GenericIssuer, operation string, err error) {
+	issuerName, nameErr := issuer.NameForIssuer(i)
+	if nameErr != nil {
+		glog.Errorf("Error recording metrics: %v", nameErr)
+	}
+
+	// count result
+	result := resultSuccess
+	if err != nil {
+		result = resultFailure
+	}
+
+	m.CertificateRequests.With(prometheus.Labels{
+		"name":        crt.Name,
+		"namespace":   crt.Namespace,
+		"issuer_name": crt.Spec.IssuerRef.Name,
+		"issuer_kind": crt.Spec.IssuerRef.Kind,
+		"issuer_type": issuerName,
+		"operation":   operation,
+		"result":      result}).Inc()
```

Comment:
I've seen other things that expose metrics (e.g. envoy) expose a distinct metric for failure events, instead of adding a `result` label to a single event type.

It'd be great to understand when and why it makes sense to do this vs many labels.
Same applies for `operation` - would we be better off defining a `renewal` counter instead of having an additional label here?

Reply:
I've added `certificate_failures_total` and `certificate_requests_total` metrics